tools/api-key-rotation/api_key_rotation_checker/main.py

```diff
@@ -0,0 +1,346 @@
+#!/usr/bin/env python3
+
+#   Copyright 2021 Google LLC
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""
+This tool checks each GCP project in your GCP Organization
+for the existence of API keys. Each key is compared
+to a custom rotation period (defaults to 90).
+
+The output consists of two groupings of API keys:
+
+- One group are keys with rotation periods older than
+the approved rotation period.
+- The second group are keys with rotation periods under
+the approved rotation period.
+"""
+
+
+import subprocess
+import requests # pylint: disable=import-error
+import googleapiclient.discovery # pylint: disable=import-error
+from datetime import datetime, timedelta
+from dateutil import parser as date_parser # pylint: disable=import-error
+from google.api_core import exceptions # pylint: disable=import-error
+from googleapiclient.errors import HttpError # pylint: disable=import-error
+from googleapiclient.discovery_cache.base import Cache # pylint: disable=import-error
+import logging
+import sys
+import argparse
+from dataclasses import dataclass
+from dataclasses_json import dataclass_json # pylint: disable=import-error
+
+
+def x_days_ago(rotation_period):
+    """
+    Get the date x days ago to use for our comparison.
+
+    Args:
+
+    rotation_period - The rotation period in days (default is 90)
+    """
+
+    rotation_days = rotation_period
+    today = datetime.now()
+
+    # Find the date x days ago
+    rotation_date = today - timedelta(days=rotation_days)
+    logging.info("%s days ago was %s \n", rotation_days, rotation_date)
+
+    return rotation_date
+
+
+def create_project_list(service):
+    """
+    Get our project list from GCP.
+
+    Args:
+
+    service - A resource manager service created via create_service() function
+    """
+
+    logging.info(
+        "Getting GCP project list. This may take a minute.. \n")
+
+    # Collect all the projects
+    projects = []
+
+    # Paginate through the list of all available projects
+    request = service.projects().list()
+    try:
+        response = request.execute()
```

Comment:
```suggestion
        response = request.execute(num_retries=3)
```
Here is example with pagination and HTTP error handling:
https://github.com/apache/airflow/blob/0f327788b5b0887c463cb83dd8f732245da96577/airflow/providers/google/cloud/hooks/gdm.py#L76-L83

Reply:
Thanks, updated